File: Others/Linear_Recurrence/Fibonacci.py
# refer to http://fusharblog.com/solving-linear-recurrence-for-programming-contest/

def do_mat_mul(mat_in1, mat_in2):
    mat_out = [[0, 0], [0, 0]]

    for r in range(2):
        for c in range(2):
            for k in range(2):
                mat_out[r][c] += (mat_in1[r][k] * mat_in2[k][c])

    return mat_out

def do_mat_pow(mat_in, p):
    if (1 == p):
        return mat_in
    elif (1 & p):
        return do_mat_mul(mat_in, do_mat_pow(mat_in, p-1))
    else:
        half = do_mat_pow(mat_in, p//2)
        return do_mat_mul(half, half)

def fib(N):
    if (N <= 0):
        return 0
    elif (N <= 2):
        return 1

    mat_trans = [[0, 1], [1, 1]]
    mat_trans = do_mat_pow(mat_trans, N-1)
    return mat_trans[0][0] + mat_trans[0][1]

print (fib(300))

File: Others/Linear_Recurrence/Fibonacci.cpp
// refer to http://fusharblog.com/solving-linear-recurrence-for-programming-contest/

#include <iostream>
#include <vector>

using namespace std;

typedef vector<vector<long long int> > matrix;

matrix mat_mul(const matrix &mat_in1, const matrix &mat_in2)
{
    const int LEN = static_cast<int>(mat_in1.size());
    matrix mat_out(LEN, vector<long long int>(LEN, 0));

    for (int r=0; r<LEN; ++r) {
        for (int c=0; c<LEN; ++c) {
            for (int p=0; p<LEN; ++p) {
                mat_out[r][c] += (mat_in1[r][p] * mat_in2[p][c]);
            }
        }
    }

    return mat_out;
}

matrix mat_pow(const matrix &mat_in, int n)
{
    if (n == 1) {
        return mat_in;
    }
    else if (n & 1) {
        return mat_mul(mat_in, mat_pow(mat_in, n-1));
    }
    else {
        const matrix half = mat_pow(mat_in, n/2);
        return mat_mul(half, half);
    }
}

int main(int argc, char** argv)
{
    int LEN = 2;
    int N = 6;

    matrix mat_trans(LEN, vector<long long int>(LEN, 0));
    mat_trans[0][1] = 1;
    mat_trans[1][0] = 1;
    mat_trans[1][1] = 1;

    if (N <= 2) {
        cout << 1 << endl;
        return 0;
    }

    mat_trans = mat_pow(mat_trans, N-1);
    cout << (mat_trans[0][0] + mat_trans[0][1]) << endl;;
    return 0;
}
